### DISCUSSION TITLE
How to handle unauthenticated user on <Link> prefetch requests

### DISCUSSION THREAD

**Discussion by joncys** [vercel/next.js#47691] [Help]
### Summary

Hello 👋🏻 .I have this handler for props on a resource that requires authentication:

```
export const getServerSideProps: GetServerSideProps<{
  // removed for brevity
}> = async ({ req, params }) => {
  const propsResult = await getUserAndReports(req)
  if (isFailure(propsResult)) {
    return {
      redirect: {
        destination: `/login?redirectTo=${req.url}`,
        permanent: false,
      },
    }
  }
  // further code
```

One can navigate to this page using the `<Link>` component within the main page of the application. But they have to be logged in.

If I open two tabs, when logged in. Log out on one of them and then click the `<Link>` in the other. Next.js tries to prefetch the props, by hitting something like `_next/data/UEb00QpF2MfGw2U37Pi4u/reports/<id>.json?id=<id>`, but since the user is not logged in, the above redirect is invoked and the user is redirected to the login page.

This is cool and works as expected.

The problem is that the `req.url` at that point is the prefetch json url (`_next/data/UEb00QpF2MfGw2U37Pi4u/reports/<id>.json?id=<id>`), so when the user completes the log in, they are redirected to the prefetch json url and not the report url they want to see.

What's the best course of action here:

- Use a RegExp to strip out the actual URL they would want to navigate to
- Return a 403 Unauthorized or something like that if the link contains `_next/data`
- Disable prefetch on the `<Link>` component
- Are there better ideas?

Thanks, all help greatly appreciated. 🫶🏻

**Comment by icyJoseph** [comment]
Have you looked into using `resolvedUrl`? https://nextjs.org/docs/api-reference/data-fetching/get-server-side-props#context-parameter

> resolvedUrl: A normalized version of the request URL that strips the _next/data prefix for client transitions and includes original query values.
